### PR TITLE
Migrate from threading.local to contextvars

### LINKED ISSUE #5036: Migrate from `threading.local` to `contextvars`
### Motivation

https://github.com/optuna/optuna/blob/3fcbf8839c760916cc0cf38fb545f526d5efe85d/optuna/study/study.py#L56 uses `threading.local`, but `contextvars` can make the code more sophisticated (https://peps.python.org/pep-0567/#converting-code-that-uses-threading-local).

### Suggestion

Use `contextvars` instead of `threading.local`.

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by SG5**
## Motivation

Migrate from threading.local to contextvars, fix #5036

## Description of the changes

Replace `_ThreadLocalStudyAttribute` class with `ContextVar` properties

### CODE REVIEW

**Comment by contramundum53** [comment]
@c-bata @HideakiImamura Could you review this PR?

**Comment by nabenabe0928** [comment]
> @c-bata @nabenabe0928 Could you review this PR?

Yes

File: optuna/study/study.py [hunk 1]
```diff
@@ -536,7 +534,7 @@ def ask(
         }
 
         # Sync storage once every trial.
-        self._thread_local.cached_all_trials = None
+        self._cached_all_trials.set(None)
```
Comment (c-bata) [inline optuna/study/study.py:537]: Could you please create a new context for each `ask()` method call instead of clearing the cache?

**Review by c-bata — CHANGES_REQUESTED** [review]

File: optuna/study/study.py (hunk 1 above)
Comment (SG5) [reply optuna/study/study.py]: Hey @c-bata,

It won't work. Object `Study` is shared between threads. Once we create a new `_cached_all_trials` in one thread, all threads lose data in context.
Test `test_cache_is_invalidated` works fine for n_jobs=1 and fails for n_jobs>1 at `assert study._cached_all_trials.get() is not None`

Another possible solution - clone `Study` for each thread

**Comment by HideakiImamura** [comment]
@SG5 Please pass the CI first. We, the reviewers, will review the PRs that have passed the CI. If you have difficulty passing the CI, it's okay to ask us for help by summarizing the error you're having trouble with and the way to reproduce it.

**Comment by SG5** [comment]
@HideakiImamura rolled back changes with recreating context, CI passed

**Comment by HideakiImamura** [comment]
@gen740 could you review this PR?

**Review by gen740 — APPROVED** [review]
LGTM!

**Comment by gen740** [comment]
It would be better to merge this PR for now. I think that creating a new context, as suggested by @c-bata, should be addressed in a follow-up PR.

**Comment by HideakiImamura** [comment]
@SG5 Please wait the merge since we, committers, will discuss the use of `ContextVar`. It will take several days.

**Comment by contramundum53** [comment]
@SG5 We are sorry, after a discussion, we found a potential bug (discrepancy from the original design) in the original `Study` class. The original intention of the issue was to use `contextvars` to simplify the caching mechanism based on the original design (which looks impossible to do now without some breaking change in the current `Study` class), and we might need to investigate the potential bug first. 
Let us close this PR for now. Thanks for working on this issue!

**Comment by contramundum53** [comment]
In the original design, we didn't allow the `Study` class to cache trials when the user directly used `ask` and `tell`. The cache is only for `optimize` function. In that case, we could create a new context for each `optimize` call.

**Pull request closed without merging**